### PR TITLE
Ban DiagnosticDescriptor ctor in CodeStyle layer

### LINKED ISSUE #50011: Ensure that analyzers in CodeStyle layer sub-type AbstractBuiltInCodeStyleDiagnosticAnalyzer or AbstractCodeQualityDiagnosticAnalyzer, not DiagnosticAnalyzer
@Youssef1313 One example: https://github.com/dotnet/roslyn/blob/64769f8b6c13abd7e931c0cf7605368da83d74c5/src/Analyzers/Core/Analyzers/RemoveUnnecessaryImports/AbstractRemoveUnnecessaryImportsDiagnosticAnalyzer.cs#L26-L27

This is turn causes IDE0005 diagnostics to not have an associated help link, falling back to bing search instead of https://docs.microsoft.com/dotnet/fundamentals/code-analysis/style-rules/ide0005: 
image

Let me file an issue and tag you, in case you are interested in fixing this up

_Originally posted by @mavasani in https://github.com/dotnet/roslyn/pull/49995#discussion_r544488364_

### PR DESCRIPTION
**Pull request by Youssef1313**
Fixes #50011

### CODE REVIEW

**Comment by Youssef1313** [comment]
@mavasani I can't figure out why the build is passing here. Can you assist please?

**Comment by mavasani** [comment]
Sure, let me investigate.

File: src/CodeStyle/BannedSymbols.txt [hunk 1]
```diff
@@ -4,4 +4,5 @@ T:Microsoft.CodeAnalysis.Options.OptionSet; 'OptionSet' is not available in Code
 T:Microsoft.CodeAnalysis.CodeStyle.CodeStyleOption`1; Use 'Microsoft.CodeAnalysis.CodeStyle.CodeStyleOption2' instead
 T:Microsoft.CodeAnalysis.CodeStyle.CodeStyleOptions; Use 'Microsoft.CodeAnalysis.CodeStyle.CodeStyleOptions2' instead
 T:Microsoft.CodeAnalysis.Options.OptionKey; Use 'Microsoft.CodeAnalysis.Options.OptionKey2' instead
-T:Microsoft.CodeAnalysis.CodeStyle.NotificationOption; Use 'Microsoft.CodeAnalysis.CodeStyle.NotificationOption2' instead
\ No newline at end of file
+T:Microsoft.CodeAnalysis.CodeStyle.NotificationOption; Use 'Microsoft.CodeAnalysis.CodeStyle.NotificationOption2' instead
+M:Microsoft.CodeAnalysis.Diagnostics.DiagnosticDescriptor.#ctor; Analyzers should extend 'AbstractBuiltInCodeStyleDiagnosticAnalyzer' or 'AbstractCodeQualityDiagnosticAnalyzer'
```
Comment (mavasani) [inline src/CodeStyle/BannedSymbols.txt:8]: You need to use the full doc comment ID (you can look at `<%repo_root%>\artifacts\bin\Microsoft.CodeAnalysis\Debug\netcoreapp3.1\Microsoft.CodeAnalysis.xml` to get this ID)
```suggestion
M:Microsoft.CodeAnalysis.DiagnosticDescriptor.#ctor(System.String,System.String,System.String,System.String,Microsoft.CodeAnalysis.DiagnosticSeverity,System.Boolean,System.String,System.String,System.String[]); Analyzers should extend 'AbstractBuiltInCodeStyleDiagnosticAnalyzer' or 'AbstractCodeQualityDiagnosticAnalyzer' instead
M:Microsoft.CodeAnalysis.DiagnosticDescriptor.#ctor(System.String,Microsoft.CodeAnalysis.LocalizableString,Microsoft.CodeAnalysis.LocalizableString,System.String,Microsoft.CodeAnalysis.DiagnosticSeverity,System.Boolean,Microsoft.CodeAnalysis.LocalizableString,System.String,System.String[]); Analyzers should extend 'AbstractBuiltInCodeStyleDiagnosticAnalyzer' or 'AbstractCodeQualityDiagnosticAnalyzer' instead
```

**Comment by mavasani** [comment]
Hopefully, you should see build failure with RS0030 diagnostics now.

**Comment by Youssef1313** [comment]
@mavasani The analyzers that cause RS0030 are run against generated code, while `AbstractBuiltInCodeStyleDiagnosticAnalyzer` doesn't.  e.g:

https://github.com/dotnet/roslyn/blob/d57cddbaa7536c8741db191be76d2a773fd0b86f/src/Analyzers/Core/Analyzers/ValidateFormatString/AbstractValidateFormatStringDiagnosticAnalyzer.cs#L71

**Comment by mavasani** [comment]
@Youssef1313 I wouldn't worry about violation in `AbstractValidateFormatStringDiagnosticAnalyzer.cs` - just add a source suppression as @Evangelink's PR https://github.com/dotnet/roslyn-analyzers/pull/4226 will soon help us remove that analyzer completely.

I think we do want to run `AbstractRemoveUnnecessaryImportsDiagnosticAnalyzer.cs` on generated code, so a source suppression with a good comment should be good for it as well.

File: src/CodeStyle/BannedSymbols.txt [hunk 2]
```diff
@@ -4,4 +4,6 @@ T:Microsoft.CodeAnalysis.Options.OptionSet; 'OptionSet' is not available in Code
...
 T:Microsoft.CodeAnalysis.Options.OptionKey; Use 'Microsoft.CodeAnalysis.Options.OptionKey2' instead
-T:Microsoft.CodeAnalysis.CodeStyle.NotificationOption; Use 'Microsoft.CodeAnalysis.CodeStyle.NotificationOption2' instead
\ No newline at end of file
+T:Microsoft.CodeAnalysis.CodeStyle.NotificationOption; Use 'Microsoft.CodeAnalysis.CodeStyle.NotificationOption2' instead
+M:Microsoft.CodeAnalysis.DiagnosticDescriptor.#ctor(System.String,System.String,System.String,System.String,Microsoft.CodeAnalysis.DiagnosticSeverity,System.Boolean,System.String,System.String,System.String[]); Analyzers should extend 'AbstractBuiltInCodeStyleDiagnosticAnalyzer' or 'AbstractCodeQualityDiagnosticAnalyzer' instead
+M:Microsoft.CodeAnalysis.DiagnosticDescriptor.#ctor(System.String,Microsoft.CodeAnalysis.LocalizableString,Microsoft.CodeAnalysis.LocalizableString,System.String,Microsoft.CodeAnalysis.DiagnosticSeverity,System.Boolean,Microsoft.CodeAnalysis.LocalizableString,System.String,System.String[]); Analyzers should extend 'AbstractBuiltInCodeStyleDiagnosticAnalyzer' or 'AbstractCodeQualityDiagnosticAnalyzer' instead
```
Comment (Youssef1313) [inline src/CodeStyle/BannedSymbols.txt:9]: Is there a way to ban *all* overloads of the constructor?

**Review by mavasani — APPROVED** [review]
LGTM

If you can also update the help links for IDE0005 descriptors, that will be nice. Thanks!

File: src/CodeStyle/BannedSymbols.txt (hunk 2 above)
Comment (mavasani) [reply src/CodeStyle/BannedSymbols.txt:9]: I don't believe so.

**Comment by mavasani** [comment]
@Youssef1313 You also have some merge conflicts.

File: src/CodeStyle/BannedSymbols.txt (hunk 2 above)
Comment (mavasani) [reply src/CodeStyle/BannedSymbols.txt:9]: @Youssef1313 That would be a feature request for the Banned API analyzer. Feel free to file one on the roslyn-analyzers repo.

**Comment by Youssef1313** [comment]
https://github.com/dotnet/roslyn/blob/d57cddbaa7536c8741db191be76d2a773fd0b86f/src/Features/Core/Portable/CodeStyle/AbstractCodeStyleProvider.cs#L77-L85

The above method is used by IDE0053

We should check whether it currently has the correct link. If it doesn't and the current build didn't fail, we need think of an alternative/enhancement. Probably add a new test to [IDEDiagnosticIDConfigurationTests] (https://github.com/dotnet/roslyn/blob/5eae4001901cb03a71c668f2aaa42783ea50570b/src/EditorFeatures/Test/Diagnostics/IDEDiagnosticIDConfigurationTests.cs) for validating the help links.

**Comment by mavasani** [comment]
> We should check whether it currently has the correct link. If it doesn't and the current build didn't fail, we need think of an alternative/enhancement

@Youssef1313 I think we are fine here - we don't expect many analyzers to be IDE-only analyzers (not present in CodeStyle layer). The guidance for future is absolutely to add analyzers to the shared layer, which implicitly adds them to CodeStyle layer. If absolutely required, we can also add the banned symbols added in this PR to https://github.com/dotnet/roslyn/blob/master/src/Features/BannedSymbols.txt

**Comment by mavasani** [comment]
@Youssef1313 Tests are still failing.

**Review by mavasani — APPROVED** [review]

**Pull request merged**